pkg/controller/operators/catalog/subscription/reconciler.go

```diff
@@ -0,0 +1,180 @@
+package subscription
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/reference"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/clientset/versioned"
+	listers "github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/listers/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/registry/reconciler"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/kubestate"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/queueinformer"
+)
+
+// catalogHealthReconciler reconciles catalog health status for subscriptions.
+type catalogHealthReconciler struct {
+	now                       func() *metav1.Time
+	client                    versioned.Interface
+	catalogLister             listers.CatalogSourceLister
+	registryReconcilerFactory reconciler.RegistryReconcilerFactory
+	globalCatalogNamespace    string
+}
+
+// Reconcile reconciles subscription catalog health conditions.
+func (c *catalogHealthReconciler) Reconcile(ctx context.Context, in kubestate.State) (out kubestate.State, err error) {
+	next := in
+	var prev kubestate.State
+
+	// loop until this state can no longer transition
+	for err == nil && out == nil && next != nil && !next.Terminal() && prev != next {
```

Comment:
shouldn't we be able to transition from one state to the same state?

Reply:
Yes, we can. The idea here is to use that as a condition for ending this reconciliation cycle.